src/include/msvc.h

```diff
@@ -34,6 +34,47 @@
 #define WT_GCC_FUNC_ATTRIBUTE(x)
 #define WT_GCC_FUNC_DECL_ATTRIBUTE(x)
 
+#define WT_UNUSED_MSVC(var) (void)(var)
+
+#define WT_ATOMIC_RELAXED 0
+#define WT_ATOMIC_ACQUIRE 1
+#define WT_ATOMIC_RELEASE 2
+#define WT_ATOMIC_ACQ_REL 3
+#define WT_ATOMIC_SEQ_CST 4
+
+#define WT_C_MEMMODEL_ATOMIC_LOAD(var, ptr, memorder) \
+    do {                                              \
+        var = *(ptr);                                 \
+        switch (memorder) {                           \
+        case WT_ATOMIC_RELAXED:                       \
+        case WT_ATOMIC_RELEASE:                       \
+            break;                                    \
+        case WT_ATOMIC_ACQUIRE:                       \
+        case WT_ATOMIC_ACQ_REL:                       \
+            WT_BARRIER();                             \
+            break;                                    \
+        case WT_ATOMIC_SEQ_CST:                       \
+            WT_FULL_BARRIER();                        \
+            break;                                    \
+        }                                             \
+    } while (0)
+
+#define WT_C_MEMMODEL_ATOMIC_STORE(ptr, value, memorder)                     \
+    do {                                                                     \
+        switch (memorder) {                                                  \
+        case WT_ATOMIC_ACQUIRE:                                              \
+        case WT_ATOMIC_ACQ_REL:                                              \
+            WT_WRITE_BARRIER();                                              \
+        case WT_ATOMIC_RELAXED:                                              \
+        case WT_ATOMIC_RELEASE:                                              \
+            *(ptr) = (value);                                                \
+            break;                                                           \
+        case WT_ATOMIC_SEQ_CST:                                              \
+            _InterlockedExchange64((volatile __int64 *)ptr, (int64_t)value); \
```

Comment:
I'm not sure whether this is required or just a write barrier is enough.

Reply:
The two common compilation strategies for seqcst store x86 are `mov mem, val; mfence;` and `xchg mem, val` (implicitly locking). With these, you need no (CPU) fencing on seq_cst reads.

Reply:
ok, does `mov mem, val; mfence;` equal to
```
*(ptr) = (value);
WT_FULL_BARRIER();
```
I thought for write, the barrier should be before the write, otherwise the compiler can still reorder writes before the barrier and writes can happen after `*(ptr) = (value);`

Does that mean we should do
```
WT_BARRIER();
*(ptr) = (value);
WT_FULL_BARRIER();
```

I guess _InterlockedExchange64 will compile to `xchg mem, val`.

Reply:
The answer to all of your questions is yes :)

x86 forbids reordering loads with loads, stores with stores and stores with earlier loads automatically. The only reordering it does allow is a store with a later load (because of store buffering and store-to-load forwarding, both of which are essentially required for performance). That is why you need a CPU fence after the store (technically you only need it at some point between the store and the next seq_cst load, but it is impractical to do better than after the store). You still need the compiler barrier though to tell the compiler before the write not to do its own reordering, since it doesn't know that the write is special.

It [looks like](https://godbolt.org/z/MWK4bK5GY) msvc, gcc, and clang have all adopted the xchg codegen, so we might as well do the same. I assume it has some advantages over the mfence version, and even if it doesn't right now, it is more likely to in the future since CPU guys optimize for what compilers generate. Plus it means we don't need to use any compiler barriers.

Reply:
Just to make sure I understand, are we discussing that we could use the full barrier just for SEQ_CST, but not for the other modes? But presumably we don't need it because we use the XCHG instruction (which has an implicit LOCK prefix)? Thanks!

Referring to my favorite website on this subject, it seems we need this only for SEQ_CST: https://www.cl.cam.ac.uk/~pes20/cpp/cpp0xmappings.html

Reply:
Yes, that's correct.